File: libs/chassis/oidc/jwks.go
package oidc

import (
	"context"
	"crypto"
	"errors"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"
)

// ErrKeyNotFound is returned when a token names a key id the authorization
// server does not serve, after a refetch has been given its chance.
var ErrKeyNotFound = errors.New("oidc: no signing key for kid")

// DefaultRefetchCooldown bounds how often an unknown key id may reach the
// network.
//
// One minute is chosen against the two failure modes either side of it. Too
// short and an unknown kid becomes a request amplifier: every caller holding a
// stale or forged token drives a fetch, and the authorization server absorbs
// our traffic. Too long and a genuine key rotation is a multi-minute outage,
// because tokens signed with the new key cannot verify until the cooldown
// lapses. Access tokens live ten minutes (§1.2), so a minute is well inside
// the window a rotation has to complete in.
const DefaultRefetchCooldown = time.Minute

// KeySet is an in-process cache of an authorization server's public signing
// keys.
//
// In-process rather than in Redis, deliberately: it is a few kilobytes that
// can always be rebuilt from the network, and putting it in Redis would add a
// hop and a failure mode to something that has neither (§15.2).
//
// The whole subtlety of this type is when it goes back to the network, and it
// is worth stating plainly because getting it wrong produces an outage that
// looks like a signature bug.
//
// A freshly seeded Zitadel serves `{"keys": []}`. It generates its signing key
// lazily, on the first token it issues, so an empty set at boot is correct
// rather than broken. A cache populated once at boot would therefore hold
// nothing for the entire life of the process and reject every token that
// followed, reporting each one as an unknown key rather than as an empty
// cache. Hence the rule this type exists to enforce: **the boot fetch must
// never be the last fetch.** Warm is explicitly not counted as a refetch, so
// the first token to arrive still gets one.
//
// Found on the real stack while building the Postman collection against a
// clean checkout, and recorded at §1.4.
type KeySet struct {
	uri      string
	client   *http.Client
	cooldown time.Duration

	// now is injectable so the refetch policy can be tested without sleeping.
	now func() time.Time

	// mu guards everything below it, and is deliberately held across the HTTP
	// fetch. That serialises concurrent misses into a single request, which is
	// the property the "exactly one refetch, not one per request" criterion
	// asks for; a lock released before the fetch would let a thundering herd
	// through. It is safe to hold because the client carries a timeout, and
	// cheap because a fetch happens at most once per cooldown.
	mu          sync.Mutex
	keys        map[string]crypto.PublicKey
	refetched   bool
	lastRefetch time.Time
}

// NewKeySet returns a cache for the JWKS served at uri. It performs no I/O;
// call Warm to populate it at boot, or let the first token drive the fetch.
func NewKeySet(uri string, client *http.Client) *KeySet {
	if client == nil {
		client = defaultClient()
	}
	return &KeySet{
		uri:      uri,
		client:   client,
		cooldown: DefaultRefetchCooldown,
		now:      time.Now,
		keys:     map[string]crypto.PublicKey{},
	}
}

// SetRefetchCooldown overrides DefaultRefetchCooldown.
func (k *KeySet) SetRefetchCooldown(d time.Duration) {
	k.mu.Lock()
	defer k.mu.Unlock()
	k.cooldown = d
}

// Warm fetches the key set once at boot.
//
// A failure here is worth logging and not worth crashing on. `auth` and
// `core-api` start together, so losing the race is ordinary, and the first
// token to arrive will fetch anyway. Crashing on it would turn a startup
// ordering detail into a restart loop.
//
// Warm does not count as a refetch. See the type comment for why that single
// line is the difference between a working stack and one that rejects every
// token it is ever shown.
func (k *KeySet) Warm(ctx context.Context) error {
	k.mu.Lock()
	defer k.mu.Unlock()
	return k.fetchLocked(ctx)
}

// KeyFor returns the public key for a key id, refetching once if the id is
// unknown.
//
// An empty kid is accepted only when the server serves exactly one key, which
// keeps this usable against an IdP that omits the header on a single-key set
// without ever weakening the check to "trust whichever key happens to work"
// (§18.2 asks for portability, not for leniency about signatures).
func (k *KeySet) KeyFor(ctx context.Context, kid string) (crypto.PublicKey, error) {
	k.mu.Lock()
	defer k.mu.Unlock()

	if key, ok := k.lookupLocked(kid); ok {
		return key, nil
	}

	if !k.mayRefetchLocked() {
		return nil, fmt.Errorf("%w: %q (refetched within the last %s)", ErrKeyNotFound, kid, k.cooldown)
	}

	// The outcome of the fetch does not change the bookkeeping: a failed or
	// fruitless refetch still starts the cooldown, or a down authorization
	// server would mean one outbound request per inbound request, which is the
	// amplification this cache exists to prevent.
	k.refetched = true
	k.lastRefetch = k.now()
	if err := k.fetchLocked(ctx); err != nil {
		return nil, fmt.Errorf("%w: %q, and the refetch failed: %v", ErrKeyNotFound, kid, err)
	}

	if key, ok := k.lookupLocked(kid); ok {
		return key, nil
	}
	return nil, fmt.Errorf("%w: %q", ErrKeyNotFound, kid)
}

func (k *KeySet) lookupLocked(kid string) (crypto.PublicKey, bool) {
	if kid != "" {
		key, ok := k.keys[kid]
		return key, ok
	}
	if len(k.keys) != 1 {
		return nil, false
	}
	for _, key := range k.keys {
		return key, true
	}
	return nil, false
}

// mayRefetchLocked answers the one question this cache exists to get right.
//
// The first miss always goes to the network, whatever Warm did, because Warm
// may have cached an empty set from an authorization server that had not
// generated its key yet. After that the cooldown applies.
func (k *KeySet) mayRefetchLocked() bool {
	if !k.refetched {
		return true
	}
	return k.now().Sub(k.lastRefetch) >= k.cooldown
}

func (k *KeySet) fetchLocked(ctx context.Context) error {
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, k.uri, nil)
	if err != nil {
		return fmt.Errorf("oidc: building jwks request: %w", err)
	}

	response, err := k.client.Do(request)
	if err != nil {
		return fmt.Errorf("oidc: fetching jwks from %s: %w", k.uri, err)
	}
	defer response.Body.Close()

	if response.StatusCode != http.StatusOK {
		return fmt.Errorf("oidc: jwks at %s returned %s", k.uri, response.Status)
	}

	body, err := io.ReadAll(io.LimitReader(response.Body, 1<<20))
	if err != nil {
		return fmt.Errorf("oidc: reading jwks: %w", err)
	}

	keys, err := parseJWKS(body)
	if err != nil {
		return err
	}
	k.keys = keys
	return nil
}

File: apps/intelligence/src/kindlast_intelligence/auth/jwks.py
"""The signing keys, fetched from the authorization server and cached.

Verification is local and in-process, never an introspection call (§1.4). That
is what makes `auth` survivable as a dependency: when it goes down, existing
tokens keep verifying until they expire, because nothing on this path talks to
it per request.
"""

from __future__ import annotations

import threading
import time
from typing import Any

import httpx
from jwt import PyJWK

from .errors import TokenInvalid

# How long to wait before refetching after an unknown `kid`.
#
# An unknown kid is the signal that the authorization server has rotated, and
# refetching is the correct response. It is also what an attacker sends to make
# this service hammer the authorization server: mint garbage with a random kid,
# repeat. The cooldown makes rotation cost one fetch and an attack cost one
# fetch, which is the same thing, which is the point.
REFETCH_COOLDOWN_SECONDS = 60.0


class KeySet:
    """A cached JWKS, refetched when a token names a key it does not hold.

    The whole subtlety of this class is when it goes back to the network, and
    it is worth stating plainly because getting it wrong produces an outage
    that looks like a signature bug.

    A FRESHLY SEEDED ZITADEL SERVES `{"keys": []}`. It generates its signing key
    lazily, on the first token it issues, so an empty set at boot is correct
    rather than broken. A cache populated once at boot would therefore hold
    nothing for the entire life of the process and reject every token that
    followed, reporting each one as an unknown key rather than as an empty
    cache. Hence the rule this class exists to enforce: **the boot fetch must
    never be the last fetch.** `warm` is explicitly not counted as a refetch,
    so the first token to arrive still gets one.

    Not hypothetical, and not cheap: the first version counted it, and the
    `Intelligence end to end` CI job failed on every run it ever executed while
    passing on a laptop every time (ENT-253). CI's stack was seconds old and
    the laptop's was a day old, so on the laptop the cooldown had long lapsed
    and the first token drove the refetch that found the key. The Go side is
    the same rule and the same reasoning, in `libs/chassis/oidc/jwks.go`.
    """

    def __init__(
        self,
        jwks_uri: str,
        client: httpx.Client | None = None,
        reachable_at: str = "",
        host_header: str = "",
    ) -> None:
        """`reachable_at` and `host_header` are the split-horizon pair.

        Discovery returns a `jwks_uri` naming the issuer's PUBLIC address,
        which in compose is not where this container can reach it. So the
        address is rewritten and the Host header carries the original.

        That changes how we get there and never who we trust: the document
        still had to declare the issuer we expected before we got this far.
        """
        if not jwks_uri:
            raise ValueError("a key set needs a jwks_uri")
        self._jwks_uri = _rewrite_host(jwks_uri, reachable_at)
        self._host_header = host_header
        self._client = client or httpx.Client(timeout=10.0)
        self._keys: dict[str, PyJWK] = {}
        # `_refetched` is separate from `_last_fetch` on purpose, and the
        # separation is the fix for ENT-253. A boot fetch updates neither, so
        # the first token to name an unknown kid always reaches the network
        # whatever `warm` did or failed to do. See the class docstring.
        self._refetched = False
        self._last_fetch = 0.0
        # Fetching and swapping the map happen under one lock. Without it two
        # requests arriving on an unknown kid both fetch, which is the
        # thundering herd the cooldown above exists to prevent.
        self._lock = threading.Lock()

    def warm(self) -> None:
        """Fetch once at boot.

        Raising here is deliberately not fatal to the caller's startup, and the
        caller is what has to honour that: an authorization server that is slow
        to come up in a compose stack should not permanently break a service
        that would have recovered on its first request. `main.build_app` logs
        and carries on for exactly that reason.

        A warm neither starts the refetch cooldown nor counts as the one
        refetch it permits. That single line is the difference between a
        working stack and one that rejects every token it is ever shown; the
        class docstring has the whole argument.
        """
        with self._lock:
            self._fetch_locked()

    def key_for(self, kid: str | None) -> PyJWK:
        with self._lock:
            key = self._keys.get(kid) if kid else self._single_key_locked()
            if key is not None:
                return key

            # Unknown kid. Refetch, but at most once per cooldown, and always
            # at least once whatever `warm` did.
            if not self._may_refetch_locked():
                raise TokenInvalid(
                    f"no signing key for kid {kid!r} "
                    f"(refetched within the last {REFETCH_COOLDOWN_SECONDS:g}s)"
                )

            # The outcome of the fetch does not change the bookkeeping. A
            # failed or fruitless refetch still starts the cooldown, or a down
            # authorization server would mean one outbound request per inbound
            # request, which is the amplification this cache exists to prevent.
            self._refetched = True
            self._last_fetch = time.monotonic()
            try:
                self._fetch_locked()
            except Exception as exc:  # noqa: BLE001 - a refusal, not a 500
                # A transport failure escaping here would leave the verifier,
                # which turns everything it catches into a refusal, and reach
                # the handler as an unhandled exception: a 500 that tells a
                # caller Intelligence is broken when the authorization server
                # is the thing that is down.
                raise TokenInvalid(
                    f"no signing key for kid {kid!r}, and the refetch failed: {exc}"
                ) from exc

            key = self._keys.get(kid) if kid else self._single_key_locked()
            if key is None:
                raise TokenInvalid(f"no signing key for kid {kid!r}")
            return key

    def _may_refetch_locked(self) -> bool:
        """The one question this cache exists to get right.

        The first miss always goes to the network, whatever `warm` did, because
        `warm` may have cached an empty set from an authorization server that
        had not generated its key yet, or may have failed outright. After that
        the cooldown applies.
        """
        if not self._refetched:
            return True
        return time.monotonic() - self._last_fetch >= REFETCH_COOLDOWN_SECONDS

    def _single_key_locked(self) -> PyJWK | None:
        """The only key, when a token names none.

        A `kid` is not required by RFC 7515, and a server serving exactly one
        key produces tokens that omit it. Serving several and omitting the kid
        is ambiguous, and guessing is how a verifier accepts a token signed by
        a key that was never meant for it, so that case is a refusal.
        """
        if len(self._keys) == 1:
            return next(iter(self._keys.values()))
        return None

    def _fetch_locked(self) -> None:
        # Deliberately does NOT touch `_refetched` or `_last_fetch`. Only
        # `key_for` records a refetch, so a boot warm cannot start the
        # cooldown. See the class docstring.
        headers = {"Host": self._host_header} if self._host_header else None
        response = self._client.get(self._jwks_uri, headers=headers)
        response.raise_for_status()
        document: dict[str, Any] = response.json()

        keys: dict[str, PyJWK] = {}
        for entry in document.get("keys", []):
            # `use: enc` is an encryption key and cannot verify a signature.
            # Skipping rather than failing: a document legitimately carries
            # both, and refusing the whole set because one key is for another
            # purpose would break a conformant server.
            if entry.get("use") == "enc":
                continue
            try:
                key = PyJWK.from_dict(entry)
            except Exception:
                # One malformed key must not drop the others. A server midway
                # through a rotation can serve something this library cannot
                # parse, and taking the whole set down over it turns a cosmetic
                # problem into an outage.
                continue
            kid = entry.get("kid")
            if kid:
                keys[kid] = key
            elif len(document.get("keys", [])) == 1:
                keys[""] = key

        # Swapped wholesale rather than merged. Merging would mean a key the
        # authorization server has retired stays valid here forever, which
        # defeats rotation.
        self._keys = keys


def _rewrite_host(url: str, reachable_at: str) -> str:
    """Point a URL at where it can actually be reached.

    ZITADEL ROUTES BY Host, AND WITHOUT THIS THE REQUEST REACHES THE WRONG
    VIRTUAL SERVER RATHER THAN FAILING CLEANLY. Measured, and the same fact
    core-api's configuration already carries: in compose, tokens name
    `http://localhost:8300` while the container reaches `http://auth:8080`, and
    asking the second for the first's documents returns a plain 404 that
    suggests the path is wrong when the path is fine.
    """
    if not reachable_at:
        return url

    from urllib.parse import urlparse, urlunparse

    target = urlparse(reachable_at)
    if not target.netloc:
        return url
    return urlunparse(urlparse(url)._replace(scheme=target.scheme, netloc=target.netloc))


def discover(
    issuer: str,
    client: httpx.Client | None = None,
    reachable_at: str = "",
    host_header: str = "",
) -> dict[str, Any]:
    """Read the OpenID configuration and check it names the issuer we asked for.

    The `jwks_uri` comes from the document rather than from an assumed path
    (§18.2), so a self-hoster can point at their own provider without a code
    change.

    The issuer check is not a formality. A document that names a different
    issuer than the one requested is either a misconfiguration or a redirect to
    somebody else's authorization server, and accepting its keys would mean
    trusting them to mint tokens for this deployment.
    """
    http = client or httpx.Client(timeout=10.0)

    # ACCEPTS AN ISSUER OR A FULL DOCUMENT URL, and appending blindly is how
    # this first failed: core-api's KINDLAST_OIDC_DISCOVERY_URL convention is
    # the complete document URL, so the container asked for
    # `/.well-known/openid-configuration/.well-known/openid-configuration` and
    # got a 404 that says nothing about the real mistake.
    #
    # Tolerating both is right rather than merely convenient: one setting names
    # the issuer and another names where to fetch its document, and a caller
    # should not have to know which half of that this function wanted.
    suffix = "/.well-known/openid-configuration"
    base = issuer.rstrip("/")
    url = base if base.endswith(suffix) else base + suffix
    expected_issuer = base[: -len(suffix)] if base.endswith(suffix) else base

    url = _rewrite_host(url, reachable_at)
    headers = {"Host": host_header} if host_header else None
    response = http.get(url, headers=headers)
    response.raise_for_status()
    document: dict[str, Any] = response.json()

    if document.get("issuer") not in (issuer, expected_issuer):
        raise ValueError(
            f"discovery document names issuer {document.get('issuer')!r}, "
            f"expected {expected_issuer!r}"
        )
    if not document.get("jwks_uri"):
        raise ValueError("discovery document has no jwks_uri")
    return document
